src/vsr/replica.zig

```diff
@@ -8357,30 +8353,58 @@ pub fn ReplicaType(
             if (header.replica == self.replica) return; // Ignore messages from self (misdirected).
 
-            const candidate: SyncTargetCandidate = switch (header.into_any()) {
-                inline .commit,
-                .ping,
-                => |h| .{
+            const candidate = switch (header.into_any()) {
+                inline .commit, .ping => |h| .{
                     .checkpoint_id = h.checkpoint_id,
                     .checkpoint_op = h.checkpoint_op,
                 },
                 else => return,
             };
 
-            if (candidate.checkpoint_op == 0) return;
-            if (candidate.checkpoint_op < self.op_checkpoint()) return;
+            if (candidate.checkpoint_op == self.op_checkpoint() and
+                candidate.checkpoint_id != self.superblock.working.checkpoint_id())
+            {
+                log.err("{}: on_{s}: jump_sync_target: checkpoint diverged " ++
+                    "(op={} id_from_superblock={x:0>32} id_from_commit={x:0>32})", .{
+                    self.replica,
+                    @tagName(header.command),
+                    candidate.checkpoint_op,
+                    self.superblock.working.checkpoint_id(),
+                    candidate.checkpoint_id,
+                });
+
+                // Either this replica, the header's replica, or both, have diverged.
+                @panic("checkpoint diverged");
+            }
+
+            // Don't sync backwards, or to our current checkpoint.
+            if (candidate.checkpoint_op <= self.op_checkpoint()) return;
 
-            if (!self.sync_target_quorum.replace(header.replica, &candidate)) {
-                if (header.command == .ping) return;
+            // Don't sync to the immediately-next checkpoint unless it has been committed atop.
+            // - If the checkpoint has been committed atop, that guarantees that at least a
+            //   commit-quorum of replicas has reached that exact checkpoint.
+            // - If the next checkpoint has *not* been committed atop yet, then we can (and should)
+            //   sync via WAL replay instead to maximize durability.
+            //
+            // To understand why this is critical, consider the case (R=3) where:
+            // 1. R₀ is primary.
+            // 2. R₀ prepares and acks all ops.
+            // 3. R₁ prepares and acks odd ops.
+            // 4. R₂ prepares and acks even ops.
+            // 5. R₀ is able to commit an entire checkpoint, but not beyond that.
+            // If R₁/R₂ state-sync to R₀'s latest checkpoint, then a single block corruption on R₀
+            // could lead to permanent unavailability.
+            if (candidate.checkpoint_op == self.op_checkpoint_next()) {
+                if (header.into_const(.commit)) |h| {
+                    if (h.commit <= self.op_checkpoint_next_trigger()) return;
+                } else {
+                    return;
+                }
```

Comment:
:thinking: it feels like we might be _not_ jumping checkpoint in some cases here. 

In particular, if a cluster is stuck in a view change, then no commit messages are flowing, and so a replica can't learn that a checkpoint that it gets from pings is commited on top of. 

Though, this is _probably_ fine --- the fact that replica lags by a checkpoint doesn't prevent it from participating in a view change, and, once view change happens, `commit` messages start flowing back.

Reply:
I guess that's one more argument for merging pings&commits

Reply:
> Though, this is probably fine --- the fact that replica lags by a checkpoint doesn't prevent it from participating in a view change, and, once view change happens, commit messages start flowing back.

Yeah, I spent some time trying to come up with a case where this doesn't work.

The closest I found was this liveness issue, which we work around in the simulator: https://github.com/tigerbeetle/tigerbeetle/pull/933#discussion_r1245440623

> I guess that's one more argument for merging pings&commits

Agreed.